reuse.go

```diff
@@ -0,0 +1,190 @@
+package libp2pquic
+
+import (
+	"errors"
+	"net"
+	"sync"
+
+	srcs "github.com/lnykww/go-src-select"
+)
+
+type ReuseConn struct {
+	net.PacketConn
+	lock sync.Mutex
+	ref  int
+}
+
+func NewReuseConn(conn net.PacketConn) *ReuseConn {
+	return &ReuseConn{
+		PacketConn: conn,
+		ref:        1,
+		lock:       sync.Mutex{},
+	}
+}
+
+func (rc *ReuseConn) Ref() error {
+	rc.lock.Lock()
+	defer rc.lock.Unlock()
+	if rc.ref == 0 {
+		return errors.New("conn closed")
+	}
+	rc.ref++
+	return nil
+}
+
+func (rc *ReuseConn) Close() error {
+	rc.lock.Lock()
+	defer rc.lock.Unlock()
+	var err error
+	switch rc.ref {
+	case 0: // cloesd, just return
+		return nil
+	case 1: // no reference, close the conn
+		err = rc.PacketConn.Close()
+	}
+	rc.ref--
+	return err
+}
+
+type Reuse struct {
+	lock           sync.Mutex
+	unicast        map[string]map[int]net.PacketConn
+	unspecific     []net.PacketConn
+	connGlobal     net.PacketConn
+	connGlobalOnce sync.Once
+}
+
+func NewReuse() *Reuse {
+	return &Reuse{
+		unicast:    make(map[string]map[int]net.PacketConn),
+		unspecific: make([]net.PacketConn, 0),
+	}
+}
+
+// getConnGlobal get the global random port socket, if not exist, create
+// it first.
+func (r *Reuse) getConnGlobal(network string) (net.PacketConn, error) {
+	var err error
+	r.connGlobalOnce.Do(func() {
+		var addr *net.UDPAddr
+		var conn net.PacketConn
+		var host string
+		switch network {
+		case "udp4":
+			host = "0.0.0.0:0"
+		case "udp6":
+			host = "[::]:0"
+		}
+		addr, err = net.ResolveUDPAddr(network, host)
+		if err != nil {
+			return
+		}
+		conn, err = net.ListenUDP(network, addr)
+		if err != nil {
+			return
+		}
+
+		r.connGlobal = NewReuseConn(conn)
+	})
+	if r.connGlobal == nil && err == nil {
+		err = errors.New("global socket init not done")
+	}
+	return r.connGlobal, err
+}
+
+// rueseConn Assertion the type of the conn is ReuseConn and inc the ref
+func (r *Reuse) reuseConn(conn net.PacketConn) error {
+	reuseConn, ok := conn.(*ReuseConn)
+	if !ok {
+		panic("type ReuseConn Assert failed: something wrong!")
+	}
+	return reuseConn.Ref()
+}
+
+func (r *Reuse) dial(network string, raddr *net.UDPAddr) (net.PacketConn, error) {
```

Comment:
We can return a `*ReuseConn` here. Then we can drop `reuseConn` and just call `Ref` directly.